lib/Trash/PageTrashBackend.php

```diff
@@ -0,0 +1,694 @@
+	 *
+	 * @return Node|null
+	 * @throws NotPermittedException
+	 */
+	private function getNodeForTrashItem(IUser $user, TrashItem $trashItem): ?Node {
+		[, $collectiveId, $path] = explode('/', $trashItem->getTrashPath(), 3);
+		if (!$this->userHasAccessToFolder($user, (int)$collectiveId)) {
+			throw new NotPermittedException('No permission to trash for collective');
+		}
+		$folders = $this->mountProvider->getFoldersForUser($user);
+		foreach ($folders as $collectiveFolder) {
+			if ($collectiveFolder['folder_id'] === (int)$collectiveId) {
+				$trashRoot = $this->getTrashFolder((int)$collectiveId);
+				try {
+					return $trashRoot->get($path);
+				} catch (NotFoundException $e) {
+					return null;
+				}
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * @return Folder
+	 * @throws NotPermittedException
+	 */
+	private function getTrashRoot(): Folder {
+		try {
+			$folder = $this->getAppFolder()->get('trash');
+			if (!$folder instanceof Folder) {
+				throw new NotPermittedException('Trash root is not a folder');
+			}
+			return $folder;
+		} catch (NotFoundException $e) {
+			return $this->getAppFolder()->newFolder('trash');
+		}
+	}
+
+	/**
+	 * @param int $collectiveId
+	 *
+	 * @return Folder
+	 * @throws NotPermittedException
+	 */
+	private function getTrashFolder(int $collectiveId): Folder {
+		try {
+			$folder = $this->getTrashRoot()->get((string)$collectiveId);
+			if (!$folder instanceof Folder) {
+				throw new NotPermittedException('Trash root is not a folder');
+			}
+			return $folder;
+		} catch (NotFoundException $e) {
+			return $this->getTrashRoot()->newFolder((string)$collectiveId);
+		}
+	}
+
+	/**
+	 * @param IUser $user
+	 * @param array $folders
+	 *
+	 * @return array
+	 * @throws NotPermittedException
+	 * @throws NotFoundException
+	 */
+	private function getTrashForFolders(IUser $user, array $folders): array {
+		$collectiveIds = array_map(static function (array $folder): int {
+			return $folder['folder_id'];
+		}, $folders);
+		$rows = $this->trashManager->listTrashForCollectives($collectiveIds);
+		$indexedRows = [];
+		foreach ($rows as $row) {
+			$key = $row['collective_id'] . '/' . $row['name'] . '/' . $row['deleted_time'];
+			$indexedRows[$key] = $row;
+		}
+		$items = [];
+		foreach ($folders as $folder) {
+			$collectiveId = $folder['folder_id'];
+			if (!$this->userHasAccessToFolder($user, (int)$collectiveId)) {
+				continue;
+			}
+
+			$mountPoint = $folder['mount_point'];
+			$trashFolder = $this->getTrashFolder($collectiveId);
+			$content = $trashFolder->getDirectoryListing();
+			foreach ($content as $item) {
+				$pathParts = pathinfo($item->getName());
+				$timestamp = (int)substr($pathParts['extension'], 1);
+				$name = $pathParts['filename'];
+				$key = $collectiveId . '/' . $name . '/' . $timestamp;
+				$originalLocation = isset($indexedRows[$key]) ? $indexedRows[$key]['original_location'] : '';
+
+				if (method_exists($item, 'getFileInfo') && null !== $info = $item->getFileInfo()) {
```

Comment:
not blocker, but in general it seems that's method that would be good to have exposed in the API, on the server, no?